---

<p align="center">
  <img src="https://raw.githubusercontent.com/lacamposm/Diplomado_Metodos_UCentral/main/data/images/imagen_ucentral.jpg" alt="logo_Ucentral" width="400px" height="300px">
</p>

# ___Diplomado: Métodos Estadísticos para el Análisis de Datos___

## ___Docente: Luis Andres Campos Maldonado.___

### ___Datos categóricos___

#### ***Ejemplo fin módulo 1.***

---

# ***SITUACIÓN PROBLEMA: TABLA DE DATOS EMBARCACIÓN TITANIC***

###  ANÁLISIS DESCRIPTIVO UNIVARIADO Y BIVARIADO

Hacia inicios del siglo 19 se buscaba realizar un viaje desde Europa hacia los EEUU, el cual lograra albergar gran cantidad de gente movilizada de continente a continente. Se presentó un desastre...
Más información [aquí](https://es.wikipedia.org/wiki/RMS_Titanic)

Se desea describir la mayor cantidad de informació en la tabla, para ello haremos uso de los conocimientos adquiridos en la parte conceptual del curso, junto al software de apoyo, en este caso, Python.


## ***INFORMACIÓN TABLA DE DATOS TITANIC.***


La información general de la tabla se encuentra consignada en el archivo
`Descripcion_tabla_titanic3.txt`. Vamos a abrir el archivo con la libreria [`request`](https://requests.readthedocs.io/en/latest/), para lograr leerlo he inspeccionar la información consignada.


In [ ]:
url_base = 'https://raw.githubusercontent.com/lacamposm/Diplomado_Metodos_UCentral/main/data/'

In [ ]:
## LECTURA DE INFORMACIÓN CONTENIDA EN LA TABLA ##
import requests

f = requests.get(url_base+"Descripcion_tabla_titanic3.txt","r")
# "r" = read
f = f.text
print(f)

NOMBRE: titanic3
TIPO: Censo
TAMAÑO: 1309 Pasajeros, 14 Variables

RESUMEN DESCRIPTIVO: La tabla de datos titanic3 describe el estado de la supervivencia
de los pasajeros en el Titanic. Los datos de titanic3
marco no contiene información para la tripulación, pero contiene
edades reales y estimadas de casi el 80% de los pasajeros.

FUENTES: Hind, Philip. "Enciclopedia Titanica". En línea. Internet.
notario público. 02 de agosto de 1999. Disponible en http://atschool.eduweb.co.uk/phind


VARIABLE DESCRIPTIONS:

pclass          Clase del pasajero.
                (1 = 1st; 2 = 2nd; 3 = 3rd).
survival        Supervivencia
                (0 = No; 1 = Yes).
name            Nombre.
sex             Sexo.
age             Edad.
sibsp           Numero de hermanos/cónyuges a bordo (esposa/prome).
parch           Numero de padres/hijos a bordo (par / hijo).
ticket          Numero del Ticket.
fare            Tarifa pagada.
cabin           Cabina.
embarked        Puerto de embarque.
                


## ***NOTEBOOK EN PYTHON.***




## LIBRERIAS UTILIZADAS

Si desea más información sobre las librerias utilizadas, puede consultarlas en los siguientes enlaces.

1. [`numpy`](https://numpy.org/doc/stable/)
2. [`pandas`](https://pandas.pydata.org/docs/)
3. [`matplotlib`](https://matplotlib.org/)
4. [`seaborn`](https://seaborn.pydata.org/)

In [ ]:
## Dataset Titanic. se puede cargar por medio de seaborn
df_sns = sns.load_dataset("titanic")
df_sns.sample(5)
## del(df_sns)
###

In [ ]:
## Data cleaning y data wrangling
import numpy as np
import pandas as pd
## Visualización
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express  as px
##
sns.set_style("darkgrid") ## Estilo para los gráficos

In [ ]:
url_base= 'https://raw.githubusercontent.com/lacamposm/Diplomado_Metodos_UCentral/main/data/'
titanic = pd.read_csv(url_base+ "titanic3.csv")
titanic.head()

pclass  survived                                             name     sex  \
0       1         1                    Allen, Miss. Elisabeth Walton  female   
1       1         1                   Allison, Master. Hudson Trevor    male   
2       1         0                     Allison, Miss. Helen Loraine  female   
3       1         0             Allison, Mr. Hudson Joshua Creighton    male   
4       1         0  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  female   

       age  sibsp  parch  ticket      fare    cabin embarked boat   body  \
0  29.0000      0      0   24160  211.3375       B5        S    2    NaN   
1   0.9167      1      2  113781  151.5500  C22 C26        S   11    NaN   
2   2.0000      1      2  113781  151.5500  C22 C26        S  NaN    NaN   
3  30.0000      1      2  113781  151.5500  C22 C26        S  NaN  135.0   
4  25.0000      1      2  113781  151.5500  C22 C26        S  NaN    NaN   

                         home.dest  
0                     St Louis, MO  
1  Montreal, PQ / Chesterville, ON  
2  Montreal, PQ / Chesterville, ON  
3  Montreal, PQ / Chesterville, ON  
4  Montreal, PQ / Chesterville, ON

Se puede observar que la tabla ha sidio cargada mis porblemas.

In [ ]:
## INFORMACIÓN DE LA TABLA CARGADA ##
titanic.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1309 entries, 0 to 1308
Data columns (total 14 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   pclass     1309 non-null   int64  
 1   survived   1309 non-null   int64  
 2   name       1309 non-null   object 
 3   sex        1309 non-null   object 
 4   age        1046 non-null   float64
 5   sibsp      1309 non-null   int64  
 6   parch      1309 non-null   int64  
 7   ticket     1309 non-null   object 
 8   fare       1308 non-null   float64
 9   cabin      295 non-null    object 
 10  embarked   1307 non-null   object 
 11  boat       486 non-null    object 
 12  body       121 non-null    float64
 13  home.dest  745 non-null    object 
dtypes: float64(3), int64(4), object(7)
memory usage: 143.3+ KB


In [ ]:
## 38% de las personas que vivieron.
titanic["survived"].mean()

0.3819709702062643

In [ ]:
titanic["survived"].value_counts()

In [ ]:
titanic.columns

Index(['pclass', 'survived', 'name', 'sex', 'age', 'sibsp', 'parch', 'ticket',
       'fare', 'cabin', 'embarked', 'boat', 'body', 'home.dest'],
      dtype='object')

Se puede observar que:
 - Las variables `pclass`, `survived`, `name`, `sex`, `sibsp`, `parch` y `ticket` tiene los registros completos, 1309 en total.
 - Las demás variables tienen registos NaN.

In [ ]:
for column in titanic.columns:
  print(f'La variable {column}, tiene ----> {sum(titanic[column].isna())} NaN.')

In [ ]:
## Conteo valores nulos por columna (campo, feature, variables)
### reemplaza el codigo de arriba en pandas
titanic.isna().sum().to_frame(name='cantidad_NaN').T

pclass  survived  name  sex  age  sibsp  parch  ticket  fare  \
cantidad_NaN       0         0     0    0  263      0      0       0     1   

              cabin  embarked  boat  body  home.dest  
cantidad_NaN   1014         2   823  1188        564

Las varibles se destribuyen así:

| **Tipo**               | **Columna**             |
|------------------------|-------------------------|
| **Continua**           | fare, age               |
| **Discreta**           | sibsp, parch            |
| **Nominal**            | survived, sex, embarked, body |
| **Ordinal**            | pclass.                   |
| **No estructurada**    | name, cabin, home.dest, boat,ticket                   |

El método `describe()` determina valores relevantes para variables númericas.

In [ ]:
## tipo de  dato: pclass y survived se deben cambiar a tipo categorico
### es de tipo object, pero realmente es categorica (en este caso no hay problema)
titanic[["pclass", "survived"]] = titanic[["pclass", "survived"]].astype("category")
titanic.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1309 entries, 0 to 1308
Data columns (total 14 columns):
 #   Column     Non-Null Count  Dtype   
---  ------     --------------  -----   
 0   pclass     1309 non-null   category
 1   survived   1309 non-null   category
 2   name       1309 non-null   object  
 3   sex        1309 non-null   object  
 4   age        1046 non-null   float64 
 5   sibsp      1309 non-null   int64   
 6   parch      1309 non-null   int64   
 7   ticket     1309 non-null   object  
 8   fare       1308 non-null   float64 
 9   cabin      295 non-null    object  
 10  embarked   1307 non-null   object  
 11  boat       486 non-null    object  
 12  body       121 non-null    float64 
 13  home.dest  745 non-null    object  
dtypes: category(2), float64(3), int64(2), object(7)
memory usage: 125.7+ KB


In [ ]:
## coluna Max
### valor acumulado
### percentiels son medidas de posicion
titanic.describe().T

count        mean        std     min      25%       50%      75%  \
age    1046.0   29.881135  14.413500  0.1667  21.0000   28.0000   39.000   
sibsp  1309.0    0.498854   1.041658  0.0000   0.0000    0.0000    1.000   
parch  1309.0    0.385027   0.865560  0.0000   0.0000    0.0000    0.000   
fare   1308.0   33.295479  51.758668  0.0000   7.8958   14.4542   31.275   
body    121.0  160.809917  97.696922  1.0000  72.0000  155.0000  256.000   

            max  
age     80.0000  
sibsp    8.0000  
parch    9.0000  
fare   512.3292  
body   328.0000

De la tabla anteior podemos observar que:

1. El 75% de las edades son de 39 años o menos.
1. hasta el 75% de las personas tiene hasta 1 un familiar
2. Dado que `survived` asigna 1 a quien sobrevivió, 38.19% de los pasajeros lo hicieron.
3. Solo el 25% tiene uno o mas `sibsp`.
4. El 75% de la muestra, no tiene `parch` entre sus familiares.
5. La varible `body` solo determina un número para el cuerpo muerto y encontrado.

In [ ]:
titanic.describe(percentiles = [0.70, 0.75, 0.80,0.85,0.9,0.95, 0.99]).T

count        mean        std     min       50%      70%      75%  \
age    1046.0   29.881135  14.413500  0.1667   28.0000   36.000   39.000   
sibsp  1309.0    0.498854   1.041658  0.0000    0.0000    1.000    1.000   
parch  1309.0    0.385027   0.865560  0.0000    0.0000    0.000    0.000   
fare   1308.0   33.295479  51.758668  0.0000   14.4542   26.955   31.275   
body    121.0  160.809917  97.696922  1.0000  155.0000  232.000  256.000   

            80%       85%        90%     95%      99%       max  
age     42.0000   45.0000   50.00000   57.00   65.000   80.0000  
sibsp    1.0000    1.0000    1.00000    2.00    5.000    8.0000  
parch    1.0000    1.0000    2.00000    2.00    4.000    9.0000  
fare    41.5792   56.9292   78.05082  133.65  262.375  512.3292  
body   269.0000  286.0000  297.00000  307.00  326.000  328.0000

Vayamos más adentro, consideremos percentiles altos y observemos que información adicional se puede extraer.

1. Solo el 5% pagó más de $ 133.65 por su boleto.
2. Solo el 5% o menos eran personas de 57 años o más.

In [ ]:
## Información relevante de las varibles tipo object
titanic.describe(include=['object']).T

count unique                   top freq
name       1309   1307  Connolly, Miss. Kate    2
sex        1309      2                  male  843
ticket     1309    929              CA. 2343   11
cabin       295    186           C23 C25 C27    6
embarked   1307      3                     S  914
boat        486     27                    13   39
home.dest   745    369          New York, NY   64

En cuanto a las varibles categóricas y no-estructuradas podemos decir:

1. En la embarcación predominaba el sexo masculino (843 de 1309).
2. La mayor cantidad de pasajeros abordaron en Souphamtom.
3. El bote 13, fue quien más pasajeros rescató, 39 en total.
4. La cuidad donde se reportaron los destino con mayor frecuencia fué New York, NY, con tan solo 64 de 1309 encuestados.
5. La información correspondiente a cabin son solo 295 de 1306 solo un poco más del 20%.

## ***Manipulación del dataframe.***


_Observe que la información que se ha subido al sistema muestra que variables como `pclass`, `survived` ingresan al sistema de tipo entero y conocemos que son variables categóricas. Además, vamos a considerar todo el problema en el idioma español_

1. _Observaremos los nombre de las columnas y usaremos un [`diccionario`](https://docs.python.org/3/tutorial/datastructures.html) para cambiarlos al idioma español y de igual forma con los valores de cada columna._
2. _Vamos a reajustar las variables a su verdadero tipo de información._
3. _Definimos 2 grupos de variables : `features_categoric` para las variables categóricas y `features_numeric` para las de tipo numérico._

In [ ]:
## Hacemos una copia del dataset original.
titanic_spa = titanic.copy()
titanic_spa.head()

pclass survived                                             name     sex  \
0      1        1                    Allen, Miss. Elisabeth Walton  female   
1      1        1                   Allison, Master. Hudson Trevor    male   
2      1        0                     Allison, Miss. Helen Loraine  female   
3      1        0             Allison, Mr. Hudson Joshua Creighton    male   
4      1        0  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  female   

       age  sibsp  parch  ticket      fare    cabin embarked boat   body  \
0  29.0000      0      0   24160  211.3375       B5        S    2    NaN   
1   0.9167      1      2  113781  151.5500  C22 C26        S   11    NaN   
2   2.0000      1      2  113781  151.5500  C22 C26        S  NaN    NaN   
3  30.0000      1      2  113781  151.5500  C22 C26        S  NaN  135.0   
4  25.0000      1      2  113781  151.5500  C22 C26        S  NaN    NaN   

                         home.dest  
0                     St Louis, MO  
1  Montreal, PQ / Chesterville, ON  
2  Montreal, PQ / Chesterville, ON  
3  Montreal, PQ / Chesterville, ON  
4  Montreal, PQ / Chesterville, ON

In [ ]:
## Nombre columnas tabla titanic3 ##
print(titanic_spa.columns.values)

['pclass' 'survived' 'name' 'sex' 'age' 'sibsp' 'parch' 'ticket' 'fare'
 'cabin' 'embarked' 'boat' 'body' 'home.dest']


In [ ]:
#### revisar 13.07.23(1:24 hrs)
## Edad en tipo categórico.
def edad_cate(value):
    """
    Categoriza la edad.
    Input: value(float)
    Output: (str) Es la categoría de la edad dependiendo del segmento
    """
    if value<18:
        return 'joven'
    elif value<50:
        return "adulto"
    elif value > 50:
        return "adulto_mayor"
    else:
      return value

##
#titanic_spa["age"].apply(edad_cate)

###apply a la columna mencionada

##Alternativa de
display(titanic_spa["age"].apply(edad_cate).head())
titanic_spa["age"].head()


NameError: ignored

In [ ]:
mapping = {"pclass":"CLASE","survived":"SUPERVIVENCIA","name":"NOMBRE","sex":"GÉNERO","age":"EDAD",
           "sibsp":"FAM_LEJA", "parch":"FAM_CERCA","ticket":"NUM_BOLETO","fare":"PRECIO_BOLETO",
           "cabin":"CABINA","embarked":"EMBARQUE","boat":"BOTE","body":"#_DE_MUERTO",
           "home.dest":"DESTINO"}
##
map_1 = {0:"MUERTO",1:"VIVO"}
map_2 = {"female":"MUJER","male":"HOMBRE"}
##
titanic_spa = titanic_spa.rename(columns=mapping)
titanic_spa["SUPERVIVENCIA"] = titanic_spa["SUPERVIVENCIA"].replace(map_1)
###replace un str
titanic_spa["GÉNERO"] = titanic_spa["GÉNERO"].replace(map_2)
titanic_spa.head()

CLASE SUPERVIVENCIA                                           NOMBRE  \
0     1          VIVO                    Allen, Miss. Elisabeth Walton   
1     1          VIVO                   Allison, Master. Hudson Trevor   
2     1        MUERTO                     Allison, Miss. Helen Loraine   
3     1        MUERTO             Allison, Mr. Hudson Joshua Creighton   
4     1        MUERTO  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)   

   GÉNERO     EDAD  FAM_LEJA  FAM_CERCA NUM_BOLETO  PRECIO_BOLETO   CABINA  \
0   MUJER  29.0000         0          0      24160       211.3375       B5   
1  HOMBRE   0.9167         1          2     113781       151.5500  C22 C26   
2   MUJER   2.0000         1          2     113781       151.5500  C22 C26   
3  HOMBRE  30.0000         1          2     113781       151.5500  C22 C26   
4   MUJER  25.0000         1          2     113781       151.5500  C22 C26   

  EMBARQUE BOTE  #_DE_MUERTO                          DESTINO  
0        S    2          NaN                     St Louis, MO  
1        S   11          NaN  Montreal, PQ / Chesterville, ON  
2        S  NaN          NaN  Montreal, PQ / Chesterville, ON  
3        S  NaN        135.0  Montreal, PQ / Chesterville, ON  
4        S  NaN          NaN  Montreal, PQ / Chesterville, ON

In [ ]:
titanic_spa.columns = [column.lower() for column in titanic_spa.columns]
titanic_spa['género'] = titanic_spa['género'].apply(lambda x: x.lower())
titanic_spa['género']

0        mujer
1       hombre
2        mujer
3       hombre
4        mujer
         ...  
1304     mujer
1305     mujer
1306    hombre
1307    hombre
1308    hombre
Name: género, Length: 1309, dtype: object

In [ ]:
#write_path = '/content/drive/MyDrive/Ucentral Maestría/Año_2023/Funda_Esta_Analítica_2023-I_GR06/prueba.csv'
#write_path1 = '/content/drive/MyDrive/Ucentral Maestría/Año_2023/Funda_Esta_Analítica_2023-I_GR06/prueba.xlsx'
#titanic_spa.to_csv(write_path, index=False)
#titanic_spa.to_excel(write_path1, index=False)


In [ ]:
titanic_spa.head()

clase supervivencia                                           nombre  \
0     1          VIVO                    Allen, Miss. Elisabeth Walton   
1     1          VIVO                   Allison, Master. Hudson Trevor   
2     1        MUERTO                     Allison, Miss. Helen Loraine   
3     1        MUERTO             Allison, Mr. Hudson Joshua Creighton   
4     1        MUERTO  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)   

   género     edad  fam_leja  fam_cerca num_boleto  precio_boleto   cabina  \
0   mujer  29.0000         0          0      24160       211.3375       B5   
1  hombre   0.9167         1          2     113781       151.5500  C22 C26   
2   mujer   2.0000         1          2     113781       151.5500  C22 C26   
3  hombre  30.0000         1          2     113781       151.5500  C22 C26   
4   mujer  25.0000         1          2     113781       151.5500  C22 C26   

  embarque bote  #_de_muerto                          destino  
0        S    2          NaN                     St Louis, MO  
1        S   11          NaN  Montreal, PQ / Chesterville, ON  
2        S  NaN          NaN  Montreal, PQ / Chesterville, ON  
3        S  NaN        135.0  Montreal, PQ / Chesterville, ON  
4        S  NaN          NaN  Montreal, PQ / Chesterville, ON

Observe que el `#_DE_MUERTO` nos proporciona como información si el cuerpo fue encontrado. Es decir, en caso que este vivo o muerto pero su cuerpo no encontrato se tendrá un NaN, en caso que el cuerpo se haya logrado encontrar se le asiga un número.

El siguiente código logra asignar 1 si el tripulante muerto fué encontrado, en cuarlquier otro caso asigna 0.

In [ ]:
titanic_spa['#_de_muerto'].isna() ## Observe esta salida.

0        True
1        True
2        True
3       False
4        True
        ...  
1304    False
1305     True
1306    False
1307     True
1308     True
Name: #_de_muerto, Length: 1309, dtype: bool

In [ ]:
titanic_spa["muerto_encontrado"] = [0 if x == True else 1 for x in titanic_spa['#_de_muerto'].isna()]
## Borrar columnas
titanic_spa.drop(columns=["#_de_muerto"], inplace = True)
##
titanic_spa

clase supervivencia                                           nombre  \
0        1          VIVO                    Allen, Miss. Elisabeth Walton   
1        1          VIVO                   Allison, Master. Hudson Trevor   
2        1        MUERTO                     Allison, Miss. Helen Loraine   
3        1        MUERTO             Allison, Mr. Hudson Joshua Creighton   
4        1        MUERTO  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)   
...    ...           ...                                              ...   
1304     3        MUERTO                             Zabour, Miss. Hileni   
1305     3        MUERTO                            Zabour, Miss. Thamine   
1306     3        MUERTO                        Zakarian, Mr. Mapriededer   
1307     3        MUERTO                              Zakarian, Mr. Ortin   
1308     3        MUERTO                               Zimmerman, Mr. Leo   

      género     edad  fam_leja  fam_cerca num_boleto  precio_boleto   cabina  \
0      mujer  29.0000         0          0      24160       211.3375       B5   
1     hombre   0.9167         1          2     113781       151.5500  C22 C26   
2      mujer   2.0000         1          2     113781       151.5500  C22 C26   
3     hombre  30.0000         1          2     113781       151.5500  C22 C26   
4      mujer  25.0000         1          2     113781       151.5500  C22 C26   
...      ...      ...       ...        ...        ...            ...      ...   
1304   mujer  14.5000         1          0       2665        14.4542      NaN   
1305   mujer      NaN         1          0       2665        14.4542      NaN   
1306  hombre  26.5000         0          0       2656         7.2250      NaN   
1307  hombre  27.0000         0          0       2670         7.2250      NaN   
1308  hombre  29.0000         0          0     315082         7.8750      NaN   

     embarque bote                          destino  muerto_encontrado  
0           S    2                     St Louis, MO                  0  
1           S   11  Montreal, PQ / Chesterville, ON                  0  
2           S  NaN  Montreal, PQ / Chesterville, ON                  0  
3           S  NaN  Montreal, PQ / Chesterville, ON                  1  
4           S  NaN  Montreal, PQ / Chesterville, ON                  0  
...       ...  ...                              ...                ...  
1304        C  NaN                              NaN                  1  
1305        C  NaN                              NaN                  0  
1306        C  NaN                              NaN                  1  
1307        C  NaN                              NaN                  0  
1308        S  NaN                              NaN                  0  

[1309 rows x 14 columns]

In [ ]:
## Seleccionamos solo las variables de tipo numéricos.
titanic_spa.info()
titanic_spa.select_dtypes(include=np.number).head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1309 entries, 0 to 1308
Data columns (total 14 columns):
 #   Column             Non-Null Count  Dtype   
---  ------             --------------  -----   
 0   clase              1309 non-null   category
 1   supervivencia      1309 non-null   category
 2   nombre             1309 non-null   object  
 3   género             1309 non-null   object  
 4   edad               1046 non-null   float64 
 5   fam_leja           1309 non-null   int64   
 6   fam_cerca          1309 non-null   int64   
 7   num_boleto         1309 non-null   object  
 8   precio_boleto      1308 non-null   float64 
 9   cabina             295 non-null    object  
 10  embarque           1307 non-null   object  
 11  bote               486 non-null    object  
 12  destino            745 non-null    object  
 13  muerto_encontrado  1309 non-null   int64   
dtypes: category(2), float64(2), int64(3), object(7)
memory usage: 125.7+ KB


edad  fam_leja  fam_cerca  precio_boleto  muerto_encontrado
0  29.0000         0          0       211.3375                  0
1   0.9167         1          2       151.5500                  0
2   2.0000         1          2       151.5500                  0
3  30.0000         1          2       151.5500                  1
4  25.0000         1          2       151.5500                  0

In [ ]:
titanic_spa.info()

In [ ]:
features_categoric = ['clase', 'supervivencia', 'género', 'embarque', 'muerto_encontrado']

In [ ]:
for i in features_categoric:
  titanic_spa[i] = titanic_spa[i].astype("category")
titanic_spa.info()

Hemos creado 2 listas, seguido hemos transformado las variables en tipo `category`, para aquellas que son variables cualitativas.

In [ ]:
print(f'{90*"-"}\n{90*"-"}\n VALORES QUE TOMAN LAS VARIABLES CATEGÓRICAS\n{90*"-"}')
##
for i in features_categoric:
   print(f'{90*"-"}\n{i} ---->{titanic_spa[i].unique()}\n')


## ***Descripción de cada una de las variables.***


### _Variables categóricas._

_La mejor manera de representar la información en este tipo de variables es una gráfica de barras, donde se puede tener en cuenta la frecuencia de cada categoría (No. de ocurrencias), o el porcentaje de participación de la misma. Es posible realizar otro tipos de gráficos, como los de pastel, aunque suelen no ser recomendados en muchas situaciones._

### ***Ejercicio 1.***

Construya una tabla de frecuencias de la variable `embarque`. Realice comentarios de los encontrado en la tabla.

In [ ]:
## Celda de código para probar.
def tabla_frecuencias(df, column=None):
  '''
      Esta función proporciona la tabla de frecuencias de una variable cualitativa.
      Input: df: pd.DataFrame,  column (str): Nombre de la columna a la cual se le busca generar la tabla de frecuencias.
      Output: pd.DataFrame con las frecuencias por categoría.
  '''
  result = df[column].value_counts().to_frame(name='frecuencia_absoluta')
  #result = result.to_frame(name='frecuencia_absoluta')
  result['frecuencia_relativa'] =  df[column].value_counts(normalize=True)
  result['frecuencia_relativa_%'] = result['frecuencia_relativa']*100
  result.loc['Totales'] = result.sum().astype('Int64')
  return result.round(2)

tabla_frecuencias(titanic_spa, 'embarque')

In [ ]:
tabla_frecuencias(titanic_spa, 'género')

### ***Ejercicio 2.***

Construya un barplot de la variable `EMBARQUE`

In [ ]:
## Celda de código para probar.
sns.countplot(data=titanic_spa, x='embarque',)
plt.xticks(rotation=0)
plt.show()

#### _Ejercicio para semana santa: Construir los demás barplot._

**Comentarios:**

1. El 69.82% embarcaron en Souphatton, lugar de salida de la embarcación.
2. Los muertos fueron 809 de 1309, es decir  61.80% aproximadamente.
3. La gran mayoría de los viajeros eran de clase 3. Cerca del 54.16% eran de dicha clase.
4. Predominaba el género masculino en la embarcación.
5. _**La variable `SUPERVIVIENCIA`, es sin duda, una de las llamativas.**_



## ***Variables cuantitativas.***


### _Variables cuantitativas._

_Debe observar que aunque en una variable se tenga una información númerica, esta no siempre se debe interpretar como tal. Para las varibles que son numéricas, es posible obtener bastante información, tanto numérica como gráfica. Entre las gráficas se destacan los histogramas, box-plot y ojivas. Así mismo medidas numéricas como las medias de tendencia central, variabilidad y de posición._

In [ ]:
titanic_numeric = titanic_spa.select_dtypes(np.number).head() ## Obtenemos un dataframe solo con estas
#                                                             ## variables, mostramos su cabeza.
titanic_numeric

In [ ]:
## Medidas númericas relevantes ##
titanic_numeric.describe()[1:].transpose()

In [ ]:
titanic_numeric.describe(percentiles=[0.80,0.85,0.9,0.95]).T

In [ ]:
titanic_spa

In [ ]:
titanic_spa.columns

In [ ]:
fig, axes = plt.subplots(2,2, figsize=(20,8))
####
g = sns.histplot(x='edad', data=titanic_spa, ax=axes[0,0] , color='blue', )
g.set_title("edad",size=15)
g.set(xlabel=" ", ylabel = "Conteo")
####
sns.histplot(x='fam_leja', data=titanic_spa, ax=axes[0,1], bins=8, color='orange' )
sns.histplot(x='fam_cerca', data=titanic_spa, ax=axes[1,0], bins=9, color='grey')
sns.histplot(x='precio_boleto', data=titanic_spa, ax = axes[1,1], bins=22, color='red' )
plt.show()

In [ ]:
fig, ax = plt.subplots(2,2,figsize=(15,6))
sns.kdeplot(x = 'EDAD', data = titanic_spa, ax = ax[0,0], common_norm = False, color ='Blue').set(title = "Kde plot")
sns.kdeplot(x = 'FAM_LEJA', data = titanic_spa, ax = ax[0,1],common_norm = True, color ='Red').set(title = "Kde plot")
sns.kdeplot(x = 'FAM_CERCA', data = titanic_spa, ax = ax[1,0],common_norm = True, color ='Orange').set(title = "Kde plot")
sns.kdeplot(x = 'PRECIO_BOLETO', data = titanic_spa, ax = ax[1,1], common_norm = True, color ='Grey').set(title = "Kde plot")
fig.tight_layout()
plt.show()

In [ ]:
fig, ax = plt.subplots(1,1,figsize = (15,6))
sns.histplot(x='edad', data=titanic_spa, color='blue', bins=16, stat="density", kde=True)
plt.show()

In [ ]:
## Box-Plots ##
#fig, ax = plt.subplots(2,2,figsize=(20,6))
#sns.boxplot(x = 'EDAD', orient = "h", data = titanic_spa, ax = ax[0,0], color = "blue")
#sns.boxplot(x = 'FAM_LEJA', data = titanic_spa, orient = "h", ax = ax[1,0], color = 'orange' )
#sns.boxplot(x = 'FAM_CERCA', data = titanic_spa, ax = ax[0,1], color = 'red')
#sns.boxplot(x = 'PRECIO_BOLETO', orient = "h", data = titanic_spa, ax = ax[1,1], color = "grey")
#plt.title('Box-plots', fontsize = 15)
#fig.tight_layout()
#plt.show()

## ***DATOS BIVARIADOS.***

_Es muy frecuente que en una investigación dos variables, ya sean ambas categóricas y/o numéricas, presenten información revelante en nuestro estudio. Vamos a analizar dichas relaciónes._

### ***TABLAS DE CONTINGENCIA.***

_Acá, vamos a considerar la relación entre dos variables categóticas, este tipo de tablas relacionan la ocurrencia de eventos simultaneos: `morir` y ser de clase alta, o la relación de `GÉNERO` con `EMBARQUE` por nombrar algunas. Para este tipo de presentaciones la librería `pandas` tiene un par de módulos: uno llamado [`crosstab`](https://pandas.pydata.org/docs/reference/api/pandas.crosstab.html) y el otro [`pivot_table`](https://pandas.pydata.org/docs/reference/api/pandas.pivot_table.html), que serán de gran utilidad._

### ***Ejercicio 3.***

Construya tres tablas de contingencias.

1. Valores totales.
2. Distribución de filas por columnas.
3. Distribución de columnas por fila.

In [ ]:
## Celda de código para probar.
pd.crosstab(titanic_spa['embarque'], titanic_spa.género)

In [ ]:
## Celda de código para probar.
pd.crosstab(titanic_spa['embarque'], titanic_spa.género, normalize='all')

In [ ]:
## Celda de código para probar.
pd.crosstab(titanic_spa['embarque'],titanic_spa.género, normalize='columns', margins=True,
            margins_name='%_zona_embarque')*100

In [ ]:
## Celda de código para probar.
pd.crosstab(titanic_spa['embarque'],titanic_spa.género, normalize='index',
            margins=True, margins_name='%_genero')*100

In [ ]:
round(pd.crosstab(titanic_spa['supervivencia'], titanic_spa['embarque'] , margins=True,
                  margins_name='% TOTAL', normalize='columns')*100,2)

In [ ]:
round(pd.crosstab(titanic_spa['género'], titanic_spa['embarque'] , normalize = "all"),4)*100

In [ ]:
round(pd.crosstab(titanic_spa['género'], titanic_spa['clase'] , normalize = "columns"),4)*100

In [ ]:
### Tabla 3.
round(pd.crosstab(titanic_spa['embarque'], titanic_spa['clase'] , normalize = "index"),4)*100

In [ ]:
## Tabla 4.
round(pd.crosstab(titanic_spa["supervivencia"], titanic_spa["embarque"] , normalize = "index"),4)*100

In [ ]:
round(pd.crosstab(titanic_spa["supervivencia"], titanic_spa["embarque"] , normalize = "columns"),4)*100

In [ ]:
## Tabla 5.
round(pd.crosstab(titanic_spa['supervivencia'], titanic_spa['género'] , margins=True,
                  margins_name= '% TOTAL', normalize = "columns")*100,2)

**Comentarios:**

1. En la variable clase, y en particular en la clase 3, se observa que hombre doblan a las mujeres.

2. En la clase 1, 61.92% sobrevivieron, en contraste en la clase 3 casi el 75% de los pasajeros murieron. En la clase 2 predomina la categoría de MUERTO, pero solo 57.04% vs. 42.96%.
3. En la tercer tabla se destaca que en su gran mayoria, en el puerto "Q", abordaron clase 3.
4. En la tabla 4 se observa que se concentra el 61.04% de los vivos en Souphanton, pero note además que también en Souphanton se encuentra el mayor % de muertos.
5. En la tabla 5 se observa que el 72.15% de los sobrevivientes son mujeres. Al parecer se le dió prioridad a salvar a las mujeres.

In [ ]:
plt.figure(figsize=(15,6))
sns.countplot(data = titanic_spa, x ="supervivencia", hue = "clase")
plt.ylabel("CONTEO")
plt.xlabel("supervivencia")
plt.title("Barplot: Supervivencia vs. Clase", fontsize = 20)
plt.show()

In [ ]:
plt.figure(figsize=(15,6))
#sns.countplot(data = titanic_spa, x = "SUPERVIVENCIA", hue = "CLASE")
#plt.ylabel("CONTEO")
#plt.xlabel("SUPERVIVENCIA")
#plt.title("Barplot: Supervivencia vs. Clase", fontsize = 20)
#plt.show()

In [ ]:

tmp = titanic_spa[titanic_spa["género"] == "mujer"]
pd.crosstab(tmp["supervivencia"],tmp["clase"], normalize = "index")

In [ ]:
tmp = titanic_spa[titanic_spa["GÉNERO"] == "MUJER"]
pd.crosstab(tmp["SUPERVIVENCIA"],tmp["CLASE"], normalize = "columns")

Observe que las gráficas refuerzan la idea anterior, la tasa de mortalidad es mucho más alta en la **clase 3** que en las demás, se observa una mayor frecuencia en los sobrevivientes de la **clase 1**. Recuerde además que la mayor cantidad de personajes en la embarcación eran de la **clase 3**.

In [ ]:
round(pd.crosstab(titanic_spa['EMBARQUE'], titanic_spa['GÉNERO'] , normalize = "columns")*100,2)

En este caso los porcentajes se obtienen por columna, es decir, el 24.35% de las mujeres embarcaron en la ciudad C, el 12.93% lo hizo en la ciudad Q y el 62.72% en la ciudad S.

De manera gráfica podemos realizar lo siguiente

In [ ]:
plt.figure(figsize=(15,6))
sns.countplot(data = titanic_spa, x = "EMBARQUE", hue = "GÉNERO")
plt.ylabel("CONTEO")
plt.xlabel("Lugar de embarcación")
plt.title("Embarcación vs. Género", fontsize = 20)
plt.show()

En la gráfica anterior vemos que aunque predomina el género masculino en todos los puertos, donde se destaca la casi paridad es el puerto Q.

Una variable de interés es sin duda `SUPERVIVENCIA`. Vamos a contrastarla con las demás categóricas.

In [ ]:
plt.figure(figsize=(15,6))
sns.countplot(data = titanic_spa, x = "SUPERVIVENCIA", hue = "GÉNERO", palette="Set2")
plt.ylabel("CONTEO")
plt.xlabel("Supervivencia")
plt.title("Supervivencia vs. Género", fontsize = 20)
plt.show()

Al parecer se dió prioridad en la salvación a las mujeres, recuerde que en la embarcación predominada el género masculino (64.4%)

In [ ]:
plt.figure(figsize=(15,6))
sns.countplot(data = titanic_spa, x = "SUPERVIVENCIA", hue = "EMBARQUE",
              palette= {"C":"red", "Q":"yellow", "S":"blue"})
plt.ylabel("Conteo")
plt.xlabel("Supervivencia")
plt.title("Supervivencia vs. Embarque", fontsize = 20)
plt.show()

Note que sobrevivieron más en el puerto C en comparación con quienes mueren. Note que en este puerto no predominaron las mujeres, para buscar atribuir esa situación dado el género. ¿Qué podría tener influencia para esta situación?

In [ ]:
##### Vamos a realizar un stack bar char en valores de % #####
import matplotlib.patches as mpatches ## Para etiquetar
## dataframes de trabajo.
total=titanic_spa.groupby('embarque')['clase'].count().reset_index() # Obtenemos un dataframe por zona de embarque y conteo por clase.
#
vivos1=titanic_spa[titanic_spa['clase'] == 1].groupby('embarque').count().reset_index()
vivos1=vivos1[['embarque','clase']] # Obtenemos un dataframe donde está el conteo por zona de embarque y la clase 1.
#
vivos2=titanic_spa[titanic_spa.clase == 2].groupby('embarque').count().reset_index()
vivos2=vivos2[['embarque','clase']] # Obtenemos un dataframe donde está el conteo por zona de embarque y la clase 2.
## Configuramos en % las clases 1 y 2
vivos1['clase']=[i/j *100 for i,j in zip(vivos1['clase'],total['clase'])]
vivos2['clase']=[i/j *100 for i,j in zip(vivos2['clase'],total['clase'])] + vivos1['clase']
total['clase']=[i/j *100 for i,j in zip(total['clase'],total['clase'])]
## Plots
plt.figure(figsize=(14,8))
sns.barplot(x='embarque', y='clase', data=total, color='darkblue')
sns.barplot(x='embarque', y='clase', data=vivos2, color='red')
sns.barplot(x='embarque', y='clase', data=vivos1, color='lightblue')
## Etiquetas
top_bar=mpatches.Patch( color='darkblue', label='clase 3')
bottom_bar=mpatches.Patch (color='red', label='clase 2')
bottom_bar1=mpatches.Patch (color='lightblue', label='clase 1')
plt.legend(handles=[bottom_bar1, bottom_bar, top_bar ], bbox_to_anchor=(1.05, 1.0), loc='upper left')
plt.title('embarque vs clase', size=20)
plt.ylabel('')
ticks=[int(x)  for x in np.linspace(0,100,11)]
labels=[str(int(x)) + '%' for x in np.linspace(0,100,11)]
plt.yticks(ticks, labels)
plt.show()

El gráfico anterior da cuenta que en el **puerto C**, predominó la **clase 1**, es decir la clase que sabemos tuvo la mayor cantidad de sobrevivientes.


## ***Variables numéricas vs. Categóricas***

Se puede analizar numérica vs cada categórica y analizar el comportamiento de
cada segmento.

Podemos determinar el % de cuerpos encontrados.

In [ ]:
titanic_spa['supervivencia'].head()

In [ ]:
(titanic_spa.query("supervivencia=='MUERTO'")["muerto_encontrado"].value_counts(normalize=True)*100).round(2)

De los fallecidos en la embarcación, solo se encontró el cuerpo del 14.96%. Realice un gráfico de la situación.

In [ ]:
pd.pivot_table(titanic_spa,                                                     ## dataframe
               values='edad',                                                   ## Columna con los valores a operar
               index = ['clase', 'género'],                                     ## Columnas de indice de fila
               columns = 'supervivencia',                                       ## Columna de índice para la columna.
               aggfunc={'edad': [np.size, np.mean, np.median, np.min, np.max,]} ## Funciones de agregación
               ).round(2)                                                       ## lambda x: np.percentile(x, 50)

Vamos a observar la distribución de la edad de aquellos que fueron sobrevivientes y aquellos que no.

In [ ]:
plt.figure(figsize=(10,8))
sns.kdeplot(x = "EDAD", hue = "SUPERVIVENCIA", data = titanic_spa, common_norm = False, cut = 0, bw_adjust=0.8)
plt.ylim(0,0.04)
plt.show()

Las distribuciones de la edad son similares entre ambas categorías, pero se observa que una mayor cantidad de niños fueron salvados en contraste con los que no.

Aunque debería ser obvio, veamos la distribución de los precios según las clases.

### ***Ejercicio 4.***

Realice un box-plot del `PRECIO_BOLETO`, con las tres clases de la embarcación. Todo en un mismo plot.

In [ ]:
## Celda de código para probar.
plt.figure(figsize=(15,6))
sns.boxplot(x="precio_boleto", # Valores en x (variable cuantitativa)
            y="clase", # Valores en y (Clases a comparar)
            data=titanic_spa, # Dataframe
            palette={1:"blue", 2:"green",3:"red"}, # Selección colores categorías.
            linewidth=0.8, # Grosor marco caja y bigotes.
            order=[3,2,1], # Selección orden de las categorías.
            fliersize=4 # Grosor outliers.
            )
plt.show()

¿Había alguna predominancia de grupo o grupos etarios dependiendo de las clases?

In [ ]:
fig, ax = plt.subplots(1,2,figsize=(20,6))
sns.kdeplot(x='edad', hue='clase', data=titanic_spa, ax=ax[0], common_norm=False, cut=0, bw_adjust=0.9).set(title="Kde plot")
sns.kdeplot(x='edad', hue='clase', data=titanic_spa, cumulative=True, common_norm=False, ax=ax[1]).set(title="Ojiva", ylabel="Densidad")
plt.show()

Observe en la **clase 1** tiene una mayor densidad en las demás en edades altas, esta clase es menos puntiaguda que las demás y tiene forma de campana, los otros 2 segmentos están sesgados a la derecha.

Al parecer, vemos que la clase era un factor importante para aumentar la chance de estar vivo, se esperaría que al comparar los precios de los boletos según si sobrevivió o no, se marcara una diferencia.

In [ ]:
## Box-Plot ##
fig, ax= plt.subplots(1,2,figsize=(20,6))
sns.kdeplot(x='PRECIO_BOLETO', hue='SUPERVIVENCIA', common_norm=False, data=titanic_spa, ax=ax[0]).set(title="Box-Plot")
sns.boxplot(x='PRECIO_BOLETO', y='SUPERVIVENCIA', orient="h", data=titanic_spa, ax=ax[1]).set(title="Box-Plot")
plt.show()

Claramente se observa la diferencia en el valor del boleto entre quienes sobrevivieron y quienes no.

In [ ]:
## Box-Plot ##
fig, ax= plt.subplots(1,1,figsize=(20,6))
sns.kdeplot(x='fam_leja', hue='supervivencia', data=titanic_spa).set(title = "Box-Plot")
plt.show()

---
## Variable numérica vs variable numérica.
---

Observemos la matriz de correlación de las variables cuantivativas.

In [ ]:
titanic_spa.select_dtypes(np.number).corr()

No se presenta una correlación alta entre ninguna variable , la correlación más descatada es estre `FAM_LEJA` y `FAM_CERCA`, veamos su comportamiento en un scatter plot.

In [ ]:
plt.figure(figsize=(15,6))
sns.scatterplot(data = titanic_spa, x = "FAM_LEJA", y = "FAM_CERCA" )
plt.show()

Otro scatterplot.

In [ ]:
plt.figure(figsize=(15,6))
sns.regplot(data = titanic_spa, x = "EDAD", y = "PRECIO_BOLETO",
            marker = "o" , # matplotlib marker code
            scatter_kws={"color": "black"}, # Color puntos
            line_kws={"color": "red"} ) # Color recta
plt.show()

In [ ]:
g = sns.FacetGrid(titanic_spa, col="CLASE",  row="SUPERVIVENCIA", height=5, aspect=1)
g.map(sns.scatterplot, "EDAD", "PRECIO_BOLETO")
g.tight_layout()
plt.show()

In [ ]:
titanic_spa.select_dtypes(np.number)

In [ ]:
g = sns.pairplot(titanic_spa.select_dtypes(np.number),diag_kind="kde", height=4, aspect=0.8,
                 plot_kws= {"color": "black"}, corner = False, diag_kws = {"color":"red"})
g.tight_layout()
plt.show()

In [ ]:
g = sns.PairGrid(titanic_spa.select_dtypes(np.number), diag_sharey=False,)
g.map_upper(sns.scatterplot, s=5, color = "red")
g.map_lower(sns.regplot, line_kws = {"color" : "black"})
g.map_diag(sns.kdeplot, lw=2, color = "green")
plt.show()

### ***Conclusión.***

Realizamos inspección, manipulación y estadísticos descriptivos sobre una tala de información, se presentaron resultados relevantes de esta tabla particular.